Fixed #26804 -- Added select_for_update() in update_or_create

jensen-cochran opened the pull request:
Added a select_for_update() when retrieving the object in update_or_create
to avoid race conditions between when the object is retrieved and when
it is saved.

timgraham commented:
Any ideas about how we could test that? Maybe `tests/select_for_update` might give some inspiration.

jensen-cochran commented:
I have been working on this and I have a few updates:

update_or_create needs to start a transaction before the .get() since select_for_update() requires being called in a transaction.  From what I can tell, the tests are all passing with this change in place because there is no test that uses update_or_create in a TransactionTestCase. In actual use the current change would throw an exception.  The tests that I need to write for this change use TransactionTestCase so that will solve that issue.

I have started writing the tests, but this somewhat difficult to test cleanly.  My current strategy is to inject some time.sleep() calls into update_or_create and run it on a new thread.  On the original thread I can do an update, which I expect to block for a period of time while update_or_create holds a lock.  After it all finishes, we would expect the data from the update on the original thread to not have been overwritten by update_or_create because I have them set to update two different fields.  With the current implementation, the test fails, but with the proposed changes (including the transaction in update_or_create), it passes.

charettes commented:
@jensen-cochran, you might want to have a look at the `select_for_update` tests which use a similar approach.

timgraham commented:
buildbot, add to whitelist.

jensen-cochran commented:
Have a test written for the case when update_or_create does an update, still need to write one for the insert case.

jensen-cochran commented:
After trying to write the test for the insert case I realized it it not needed--if the row doesn't exist there is nothing to lock.  So I think the test I have should be enough.

If this gets pulled in I will make sure to submit the CLA asap!

timgraham commented on tests/get_or_create/tests.py:
```diff
@@ -413,3 +417,56 @@ def test_update_callable_default(self):
+    @skipUnlessDBFeature('supports_transactions')
+    def test_updates_in_transaction(self):
+        """
+        Should select and update objects in a transaction to avoid race conditions. This test basically forces
```
Please wrap docstrings at 79 characters. Ticket references are only needed for obscure issues that benefit from the additional context of the ticket. Not sure that's the case here.

timgraham commented:
Can you think of any backwards-compatibilities that users might see which could be worth mentioning in the release notes? You may squash commits when updating, thanks!

jensen-cochran commented:
Fixed the docstrings and squashed.

In #26638, the ability was added to pass callables into defaults for get_or_create and update_or_create.  I was thinking that it may be useful to warn users that since update_or_create now locks the row that they should not try to do too much work inside those callables because it will lock the row for longer, but I ran into an issue.

While I was playing with passing callables into defaults to lock the row for a while before update, I realized that update_or_create does not evaluate callables in defaults when update_or_create is updating, only when it is creating.  When updating, the function will be cast to a string and Django will try to store that into the database.  There is one update_or_create test added in #26638, but it only tests the insert case, not the update case.

I think that #26638 should be reopened or a new issue should be made (whatever is standard, I'm new here) to address this.  It would also make my test for this issue cleaner because I could pass a callable with a  sleep() into defaults instead of what I am doing now.

timgraham commented:
We'd open a new ticket since the commits in the other ticket are already released.

timgraham commented:
Ready for a rebase and test update, following 841cb45847f9a96c108e038300006b4c45aef12a.

jensen-cochran commented:
Rebased and updated test to use callable

timgraham commented:
merged in d44afd889275473c97474cca19467d1509e0fcc1, thanks!

timgraham commented:
Unfortunately, I've seen this new time-dependent test fail more than once on Oracle. e.g.
```
  File "/home/jenkins/workspace/django-oracle/database/oracle12/label/trusty/python/python3.4/tests/get_or_create/tests.py", line 472, in test_updates_in_transaction
    self.assertGreater(after_update - before_start, timedelta(seconds=0.3))
AssertionError: datetime.timedelta(0, 0, 61485) not greater than datetime.timedelta(0, 0, 300000)
```

jensen-cochran commented:
One of two things have to be happening: either Oracle is not locking the row, or it is taking longer than I anticipated to lock the row. The 50ms sleep I have in there is meant to make sure that the database has time to grab the lock, but maybe that is not enough in all cases.  

I think I have a way to fix this. I can have birthday_sleep() set a flag to signal that the lock has been grabbed and that the thread is about to sleep.  I'll just poll that flag in the main thread.

jensen-cochran commented:
I have made an issue for this [#26933](https://code.djangoproject.com/ticket/26933)
